### PR TITLE
pnap machine driver implementation

### PR DESCRIPTION
**Pull request by pajuga**
### Summary
phoenixNAP (pnap) is Rancher node driver which has its own cloud credential type in Rancher. Node provisioning and cluster creation works as expected when RKE1 is used as provisioner. This PR should solve issue to provision new node (through phoenixNAP node driver) and create cluster using RKE2/K3s. Additionally, implementation to dynamically pull locations, types, os is added as part of this PR.

fixes rancher/dashboard#6695

### Technical notes summary

Added cloud credentials, machine config, model, store components for pnap machine driver.

### Areas or cases that should be tested

Steps to reproduce: 
Go to Cluster Management, 
Clusters -> Create, 
Provision new nodes and create a cluster using RKE2/K3s -> select phoenixNAP, 
Select pnap cloud credentials. Verify that there is custom UI for pnap which dynamically populates Location Type and OS drop downs, and shows details about selected type.
Click on create button to verify that cluster is successfully created if valid credentials are entered.

### CODE REVIEW

**Comment by richard-cox** [comment]
@pajuga Please can you fill out the PR template below. Could you include some background on what pnap is and how we can validate these changes?



### Summary
Fixes #


### Occurred changes and/or fixed issues


### Technical notes summary


### Areas or cases that should be tested



### Areas which could experience regressions


### Screenshot/Video

**Comment by pajuga** [comment]
@richard-cox I have added some background about this PR, please let me know if you need more details.
Also, wanted to check something about sending cloud credentials over to node driver.
Cloud credentials get created with new UI successfully and can be used by rke1 provisioner without any issues.
Pnap driver [data](https://github.com/rancher/rancher/blob/1e9a6b0ddcba29658a214700a073747702d7f832/pkg/data/management/machinedriver_data.go#L49) have public credential field clientIdentifier 
and private credential field  clientSecret.
 
Thing is that clientSecret is sent to node driver (docker-machine-driver) from new UI (from machine pools) with RKE2 provisioner without any additional customization, but only way I found to send clientIdentifier from selected cloud credentials to node driver was to set this [line](https://github.com/PNAP/dashboard/blob/060de9f8b545c8dcd2b48d38b7efc8060d2f9cc4/shell/machine-config/pnap.vue#L41) in UI code.
I am not happy with this implementation mostly because if Client ID gets updated, change will not be recognized in node driver (docker-machine-driver) until machine pool gets updated too. This problem does not happen in case of Client Secret, when secret gets updated change is sent to node driver without any additional actions.

File: shell/machine-config/pnap.vue [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        for (const prod of products) {
+          for (const plan of prod.plans) {
+            if (
+              !map.has(prod.productCode) &&
+              plan.location === serverLocation
+            ) {
+              map.set(prod.productCode, true);
+              result.push({
+                value: prod.productCode,
+                label: `${ prod.productCode } - $${ plan.price }/hour`,
```
Comment (richard-cox) [inline shell/machine-config/pnap.vue:125]: No change required, just confirmation, pnap only charge in USD correct?

File: shell/machine-config/pnap.vue [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    <div class="row mt-20">
+      <div class="col span-6">
+        <LabeledSelect
+          v-model="value.serverLocation"
+          :mode="mode"
+          :options="locationChoices"
+          :searchable="true"
+          :required="true"
+          :disabled="disabled"
+          label="Location"
```
Comment (richard-cox) [inline shell/machine-config/pnap.vue:192]: These need localising. I've added it to this line and the pattern should reproducible

**Review by richard-cox — CHANGES_REQUESTED** [review]
PR looks really good! I've played around with pnap provisioning and creds. Only a few minor comments, plus I've pushed a commit with some minor tweaks
- Credentials form layout
  - From
    image
  - To
     image
- Ensure cpu, cores, etc fields are shown as disabled rather than just read-only
- Add example of l10n for one of the form fields

In regards to your question https://github.com/rancher/dashboard/pull/6569#issuecomment-1209611616... good question. I haven't dived in to `rancher/rancher` machine config code though, so I'm no best placed to answer. If you have a contact their side it would be better to ask them, if you don't let me know and I'll find one.


Update - I couldn't push my changes to your repo/branch. Patch is below

```
From 68269218bf5690eef6c8279e0d6e090ce97f6731 Mon Sep 17 00:00:00 2001
From: Richard Cox <richard.cox@suse.com>
Date: Fri, 19 Aug 2022 14:34:51 +0100
Subject: [PATCH] Minor tweaks - Add better spacing in the credentials form -
 Ensure cpu, cores, etc fields are shown as disabled rather than just
 read-only - Add example of l10n for one of the form fields

---
 shell/assets/translations/en-us.yaml |  3 ++
 shell/cloud-credential/pnap.vue      | 42 +++++++++++++++++-----------
 shell/machine-config/pnap.vue        | 16 +++++------
 3 files changed, 36 insertions(+), 25 deletions(-)

diff --git a/shell/assets/translations/en-us.yaml b/shell/assets/translations/en-us.yaml
index cc1c7efae..cd3486cf6 100644
--- a/shell/assets/translations/en-us.yaml
+++ b/shell/assets/translations/en-us.yaml
@@ -1441,6 +1441,9 @@ cluster:
       httpEndpoint: Allow access to EC2 metadata
       httpTokens: Use tokens for metadata
       tagTitle: EC2 Tags
+    pnap:
+      serverLocation:
+        label: Location
   addOns:
     dependencyBanner: Add-On Configurations can vary between Kubernetes versions. Changing the Kubernetes version may reset the values below.
     additionalManifest:
diff --git a/shell/cloud-credential/pnap.vue b/shell/cloud-credential/pnap.vue
index 714bd70a1..c1f59e912 100644
--- a/shell/cloud-credential/pnap.vue
+++ b/shell/cloud-credential/pnap.vue
@@ -25,22 +25,30 @@ export default {
 
 <template>
   <div>
-    <LabeledInput
-      :value="value.decodedData.clientIdentifier"
-      label-key="cluster.credential.pnap.clientIdentifier.label"
-      placeholder-key="cluster.credential.pnap.clientIdentifier.placeholder"
-      type="text"
-      :mode="mode"
-      @input="value.setData('clientIdentifier', $event);"
-    />
-    <LabeledInput
-      :value="value.decodedData.clientSecret"
-      label-key="cluster.credential.pnap.clientSecret.label"
-      placeholder-key="cluster.credential.pnap.clientSecret.placeholder"
-      type="text"
-      :mode="mode"
-      @input="value.setData('clientSecret', $event);"
-    />
-    <p class="text-muted mt-10" v-html="t('cluster.credential.pnap.clientSecret.help', {}, true)" />
+    <div class="row mt-20">
+      <div class="col span-6">
+        <LabeledInput
+          :value="value.decodedData.clientIdentifier"
+          label-key="cluster.credential.pnap.clientIdentifier.label"
+          placeholder-key="cluster.credential.pnap.clientIdentifier.placeholder"
+          type="text"
+          :mode="mode"
+          @input="value.setData('clientIdentifier', $event);"
+        />
+      </div>
+      <div class="col span-6">
+        <LabeledInput
+          :value="value.decodedData.clientSecret"
+          label-key="cluster.credential.pnap.clientSecret.label"
+          placeholder-key="cluster.credential.pnap.clientSecret.placeholder"
+          type="text"
+          :mode="mode"
+          @input="value.setData('clientSecret', $event);"
+        />
+      </div>
+    </div>
+    <div class="row mt-5">
+      <p class="text-muted mt-10" v-html="t('cluster.credential.pnap.clientSecret.help', {}, true)" />
+    </div>
   </div>
 </template>
diff --git a/shell/machine-config/pnap.vue b/shell/machine-config/pnap.vue
index d75754206..6e29bc235 100644
--- a/shell/machine-config/pnap.vue
+++ b/shell/machine-config/pnap.vue
@@ -189,7 +189,7 @@ export default {
           :searchable="true"
           :required="true"
           :disabled="disabled"
-          label="Location"
+          :label="t('cluster.machineConfig.pnap.serverLocation.label')"
         />
       </div>
       <div class="col span-6">
@@ -209,7 +209,7 @@ export default {
         <LabeledInput
           v-model="type.metadata.cpu"
           :mode="mode"
-          :readonly="true"
+          :disabled="true"
           label="CPU"
         />
       </div>
@@ -217,7 +217,7 @@ export default {
         <LabeledInput
           v-model="type.metadata.coresPerCpu"
           :mode="mode"
-          :readonly="true"
+          :disabled="true"
           label="Cores per CPU"
         />
       </div>
@@ -225,7 +225,7 @@ export default {
         <LabeledInput
           v-model="type.metadata.cpuCount"
           :mode="mode"
-          :readonly="true"
+          :disabled="true"
           label="CPU Count"
         />
       </div>
@@ -233,7 +233,7 @@ export default {
         <LabeledInput
           v-model="type.metadata.cpuFrequency"
           :mode="mode"
-          :readonly="true"
+          :disabled="true"
           label="CPU Frequency [GHz]"
         />
       </div>
@@ -243,7 +243,7 @@ export default {
         <LabeledInput
           v-model="type.metadata.ramInGb"
           :mode="mode"
-          :readonly="true"
+          :disabled="true"
           label="RAM [GB]"
         />
       </div>
@@ -251,7 +251,7 @@ export default {
         <LabeledInput
           v-model="type.metadata.storage"
           :mode="mode"
-          :readonly="true"
+          :disabled="true"
           label="Storage"
         />
       </div>
@@ -259,7 +259,7 @@ export default {
         <LabeledInput
           v-model="type.metadata.network"
           :mode="mode"
-          :readonly="true"
+          :disabled="true"
           label="Network"
         />
       </div>
-- 
2.35.3

```

File: shell/cloud-credential/pnap.vue [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      this.$emit('validationChanged', !!neu);
+    },
+    'value.decodedData.clientSecret'(neu) {
+      this.$emit('validationChanged', !!neu);
+    }
+  },
+
+  methods: {
+    test() {
+      return true;
```
[the remark this answers is not in the document]
Comment (pajuga) [reply shell/cloud-credential/pnap.vue]: I added http request to check credentials. Yes these two are public. It is allowed to see product inventory through API, even without account.

File: shell/machine-config/pnap.vue (hunk 1 above)
Comment (pajuga) [reply shell/machine-config/pnap.vue:125]: Pnap can charge EU customers in euros, but all prices are in dollars and just converted.

File: shell/machine-config/pnap.vue (hunk 2 above)
Comment (pajuga) [reply shell/machine-config/pnap.vue]: These have been localized.

**Comment by pajuga** [comment]
@richard-cox Thank you for reviewing and attaching patch. Your patch is applied and I believe I resolved all other things. Could you please review. 
In regards to your question about [comment](https://github.com/rancher/dashboard/pull/6569#issuecomment-1209611616) I was in communication with @snasovich about rancher/rancher repo. Sergey was not sure if problem is on UI or backend so I asked   this here.

**Comment by pajuga** [comment]
> Can you provide some more detail, what is currently happening and what you expect to happen? Is the information being sent correctly? Please contain as much detail about the http requests, resource types you're modifying, and what you would expect to see when fetching them.

S

Scenario A
1 Create pnap (rke2) cluster with selected cloud credentials and one pool is enough.
2 Change/update client secret of cloud credentials that are used for cluster.
3 Scale pool up
Current behavior: new client secret is used for provisioning of new node.
Expected behavior is same as current behavior

Scenario B
1 Create pnap (rke2) cluster with selected cloud credentials and one pool is enough.
2 Change/update client id of cloud credentials that are used for cluster.
3 Scale pool up
Current behavior: old client id is still used for provisioning of new node.
Expected behavior: new (updated) client id is used for provisioning of new node.

New client id will be used after cluster config is saved again even without any changes.

Is this supposed to work this way or there is some issue.

**Comment by richard-cox** [comment]
Thank you for the clarification

> Is this supposed to work this way or there is some issue, it seems information about client id is not properly sent from dashboard ui to the rancher backend and pnap node driver in case of scenario A.

Could you confirm that the request's that the browser makes (via DevTools network tab) contains the correct information?

**Comment by pajuga** [comment]
After i edit cloud credentials and click on save button i see two requests:
1. http request to check credentials. Contains correct information and passed 
2. http request v3/cloudCredentials/cattle-global-data:cc-zzfc5 contains payload with pnapcredentialConfig
and correct information about clientIdentifier and clientSecret

After i click scale pool up button I see following http POST request v1/provisioning.cattle.io.clusters/fleet-default/b
b is name of the pool and request payload has spec with cloudCredentialSecretName: "cattle-global-data:cc-zzfc5".

At some point node driver will start to provision new node for the pool and credentials that are received by node driver are clientIdentifier before cloud credentials update and proper value of cloudSecret (value after updafe).

So it seems to me all requests send correct information to the backend, but somehow node pool does not pull value of clientIdentifier from the cloud credentials that are saved on backend. Node pool has correct cloud credentials ID but when it comes to the part to take value of clientIdentifier and send it to node driver, that value is not properly read from cloud credentials saved on backend. Interesting, this happens only for clientIdentifier, clientSecret works as expected.

**Comment by richard-cox** [comment]
Thanks for the update @pajuga . It looks like that isn't a UI/dashboard issue though, so someone from outside of our team will need to take a look. Did you say you had another contact within Rancher or I could find someone to help?

**Comment by richard-cox** [comment]
@snasovich from my understanding

1) Cluster is provisioned with cloud credentials 
2) Cloud credentials are successfully updated
3) Scaling of nodes in cluster fail due to incorrect cloud credentials

- Is this a valid use case, can cloud credentials be updated and used by an existing cluster?
- It looks like the correct requests are made to update the credentials with the correct content, are those the right one's to make?

**Comment by snasovich** [comment]
@pajuga @richard-cox , apologies for late response here.
On cluster YAML, is the only value that refers to cloud credentials `cloudCredentialSecretName` which then has the value like `cattle-global-data:cc-xxxxx` which is a k8s secret that contains both Client ID and Client Secret values for pnap account access?
Or is there some other field on cluster spec that refers to the Client ID?

If it's the former, this is something that will require a deeper investigation as it would be really strange why of 2 values contained on cloud credential secret only one is updated on the scale up request. If there was some sort of caching, I would expect both to have old values.

**Comment by pajuga** [comment]
@snasovich @richard-cox Apologies, I just realized I have never replied back. I think cloudCredentialSecretName is the only field that refers to cloud credentials in cluster configuration.

Guys does it make sense to merge this PR and create separate issue for this specific scenario. Other good things are blocked with this scenario which seems to happen rarely.

**Review by richard-cox — COMMENTED** [review]
@pajuga Agree, lets get this in. Can you confirm the functionality still works and if so i'll approve and merge

- Create a PNAP cluster
- Can view the cluster (Burger menu --> `Cluster Management` click on the cluster), 
  - Detail tab - everything looks fine and can use machine pool buttons available
  - Config tab - Machine Pool fields are disabled
- Can edit the cluster 
  - Change any of the applicable machine config

You will need to supply @thaneunsoo with some credentials to verify the changes. I'll leave you two to connect, but it's probably easiest in `rancher-users` slack

**Comment by pajuga** [comment]
@thaneunsoo I have contacted you on slack as @richard-cox suggested. Could you please check if you get the message.

**Comment by thaneunsoo** [comment]
I am getting a sha256 checksum error. The cluster is unable to download the driver correctly because of the checksum not matching up.

**Comment by richard-cox** [comment]
@thaneunsoo This hasn't merged yet, just trying to get all ducks aligned before proceeding.

@pajuga could you confirm that with this PR the flow in https://github.com/rancher/dashboard/pull/6569#pullrequestreview-1221130265 works as expected? We're approaching code complete for the next release at the end of this week and it would be good to get this in by then.

**Comment by richard-cox** [comment]
> I am getting a sha256 checksum error. The cluster is unable to download the driver correctly because of the checksum not matching up.
> 
> 

@pajuga  Does the `Download URL` / `Checksum` of the pnap node driver need updating?

I'm my 2.7-next1 system it's down as 
- https://github.com/phoenixnap/docker-machine-driver-pnap/releases/download/v0.4.0/docker-machine-driver-pnap_0.4.0_linux_amd64.zip
- 0bc81bdc80ab258fa0db67918f3b04435ed2c81f84c942c9123a0729f884190b

**Comment by pajuga** [comment]
> > I am getting a sha256 checksum error. The cluster is unable to download the driver correctly because of the checksum not matching up.
> > 
> 
> @pajuga Does the `Download URL` / `Checksum` of the pnap node driver need updating?
> 
> I'm my 2.7-next1 system it's down as
> 
> * https://github.com/phoenixnap/docker-machine-driver-pnap/releases/download/v0.4.0/docker-machine-driver-pnap_0.4.0_linux_amd64.zip
> * 0bc81bdc80ab258fa0db67918f3b04435ed2c81f84c942c9123a0729f884190b

Hi @richard-cox, I tested this with Rancher 2.6.0, and found no issues. That is the most recent version of driver and checksum.
Also i confirmed workflow works as expected. I will try 2.7.0 too

**Comment by pajuga** [comment]
@richard-cox  @thaneunsoo I tested on 2.7.0 and 2.7-head versions and did not notice any problem with pnap driver download and checksums. I managed to activate and use pnap node driver without any issues. Could you please write me steps to reproduce the problem.

**Review by richard-cox — APPROVED** [review]

**Comment by richard-cox** [comment]
@pajuga No specific bug, just wanted to confirm that there was no bit rot and the successful case worked before we pass to QA

**Comment by richard-cox** [comment]
@pajuga There's been some billing issues with the account used to test this. I think we set it up after chatting in `rancher-users.slack.com` slack, could you ping me there?

**Pull request merged**